### PR TITLE
Add enforcement of sending signals to arbitrary processes in a container

### PR DESCRIPTION
**Pull request by SeanTAllen**
Most people tend to think of a container as thing that does a single thing
and by extension, as a thing that is a single process. This isn't the case
but it is the mental model that people bring to containers. It also happens
to be an incredibly common scenario where a container "equals" a single process.

Our model for representing "container constraints" in the policy engine in GCS
doesn't break out the init process of a container into a process representation.
Instead, signals going to an init process are represented as signals "on the
container" and signals going to other processes in the container are represented
as signals on an "exec_process".

This makes understanding how to apply constraints to a container easier for
folks getting started with confidential containers. It also means that we
need to have two different bits of code for enforcing signal constraints on
processes in GCS.

For any pid and container id combo we do the following:

If the pid is for the init process of a container, then we know to use the
container signal list for determining if sending the signal is allowed.

If the pid isn't for an init process, then we have a bit more work to do.
We don't have any mapping of a pid to policy entry if the pid isn't for
the init process of a container. But, we can still map from the incoming
pid to set of 0 or more valid policy exec_process entries.

GCS stores a collection, on a per container basis, of all processes that were
started including the OCI Spec for each process started. We can match the
pid to the OCI Spec used to start the process with said pid. We can then use
the command from that process to find any valid exec process entries in the
policy for the container and see if any of them allow the signal in question.
Any matching exec process entries can then be used to do further narrowing on
possible container matches in the policy for the given container id.

Signed-off-by: Sean T. Allen <seanallen@microsoft.com>

### CODE REVIEW

**Comment by SeanTAllen** [comment]
@matajoh this is post rebase. i reviewed the PR. it looks good to me now. but, it is missing error messaging in framework.rego and I'm not sure the correct way to tackle it.

File: pkg/securitypolicy/regopolicy_test.go [hunk 1]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+				return false
+			}
+		}
+
+		signal := selectSignalFromSignals(testRand, hasAllowedSignals.Signals)
+		badContainerID := generateContainerID(testRand)
+		err = tc.policy.EnforceSignalContainerProcessPolicy(badContainerID, signal, true, hasAllowedSignals.Command)
+
+		if err == nil {
+			t.Errorf("Signal init process expectedly succeeded: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1445]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 2]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+				return false
+			}
+			containerID = r.containerID
+		}
+
+		signal := selectSignalFromSignals(testRand, hasAllowedSignals.Signals)
+		err = tc.policy.EnforceSignalContainerProcessPolicy(containerID, signal, true, hasAllowedSignals.Command)
+
+		if err != nil {
+			t.Errorf("Signal init process expectedly failed: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1368]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 3]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("Unable to exec process for test: %v", err)
+			return false
+		}
+
+		signal := selectSignalFromSignals(testRand, processUnderTest.Signals)
+
+		err = tc.policy.EnforceSignalContainerProcessPolicy(containerID, signal, false, processUnderTest.Command)
+		if err != nil {
+			t.Errorf("Signal init process expectedly failed: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1496]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 4]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("Unable to exec process for test: %v", err)
+			return false
+		}
+
+		signal := generateSignal(testRand)
+
+		err = tc.policy.EnforceSignalContainerProcessPolicy(containerID, signal, false, processUnderTest.Command)
+		if err == nil {
+			t.Errorf("Signal init process expectedly succeeded: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1547]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 5]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+			t.Errorf("Unable to exec process for test: %v", err)
+			return false
+		}
+
+		signal := selectSignalFromSignals(testRand, processUnderTest.Signals)
+		badCommand := generateCommand(testRand)
+
+		err = tc.policy.EnforceSignalContainerProcessPolicy(containerID, signal, false, badCommand)
+		if err == nil {
+			t.Errorf("Signal init process expectedly succeeded: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1599]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 6]
```diff
@@ -1337,6 +1339,327 @@ func Test_Rego_ShutdownContainerPolicy_Not_Running_Container(t *testing.T) {
...
+			t.Errorf("Unable to exec process for test: %v", err)
+			return false
+		}
+
+		signal := selectSignalFromSignals(testRand, processUnderTest.Signals)
+		badContainerID := generateContainerID(testRand)
+
+		err = tc.policy.EnforceSignalContainerProcessPolicy(badContainerID, signal, false, processUnderTest.Command)
+		if err == nil {
+			t.Errorf("Signal init process expectedly succeeded: %v", err)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1651]: nit: unexpectedly

File: pkg/securitypolicy/regopolicy_test.go [hunk 7]
```diff
@@ -1521,42 +1861,53 @@ func setupRegoRunningContainerTest(gc *generatedConstraints) (tc *regoRunningCon
 	numOfRunningContainers := atLeastOneAtMost(testRand, int32(len(gc.containers)))
 	for i := 0; i < int(numOfRunningContainers); i++ {
 		containerToStart := gc.containers[0]
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1863]: Just noticing this, do you mean to start the same container each time?

**Review by matajoh — CHANGES_REQUESTED** [review]
Just a few nits around the error messages.

**Review by matajoh — APPROVED** [review]

**Pull request merged**